src/core/ddsc/tests/deadline.c

```diff
@@ -434,3 +437,117 @@ CU_Theory((int32_t n_inst, uint8_t unreg_nth, uint8_t dispose_nth), ddsc_deadlin
     }
   } while (!test_finished);
 }
+
+#define DEADLINE DDS_MSECS(500)
+
+static void cb (struct ddsi_xevent *xev, void *ptr, ddsrt_mtime_t tm)
+{
+  (void) xev;
+  (void) ptr;
+  (void) tm;
+  //this function's purpose is to delay the monitor thread such that while instance's deadline may expire, but their updates are "queued" if updates happen while the monitor thread is "busy"
+  dds_sleepfor(DEADLINE);
+}
+
+static void check_statuses(dds_entity_t datareader, dds_entity_t datawriter, uint32_t prevcount, uint32_t newcount, dds_instance_handle_t prevhandle, dds_instance_handle_t newhandle)
+{
+  dds_requested_deadline_missed_status_t rstatus;
+
+  dds_return_t rc = dds_get_requested_deadline_missed_status (datareader, &rstatus);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  CU_ASSERT ( (rstatus.total_count == prevcount && rstatus.last_instance_handle == prevhandle) ||
+              (rstatus.total_count == newcount  && rstatus.last_instance_handle == newhandle));
+
+  dds_offered_deadline_missed_status_t ostatus;
+
+  rc = dds_get_offered_deadline_missed_status (datawriter, &ostatus);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  CU_ASSERT ( (ostatus.total_count == prevcount && ostatus.last_instance_handle == prevhandle) ||
+              (ostatus.total_count == newcount  && ostatus.last_instance_handle == newhandle));
+}
+
+CU_Test(ddsc_deadline, update)
+{
+  dds_entity_t pp = dds_create_participant(DDS_DOMAIN_DEFAULT, NULL, NULL);
+  CU_ASSERT_FATAL(pp > 0);
+
+  dds_entity_t tp = dds_create_topic(pp, &Space_Type1_desc, "abc_def", NULL, NULL);
+  CU_ASSERT_FATAL(tp > 0);
+
+  //qos
+  dds_qos_t qos;
+  ddsi_xqos_init_empty (&qos);
+  qos.present |= DDSI_QP_HISTORY | DDSI_QP_DESTINATION_ORDER | DDSI_QP_DEADLINE;
+  qos.history.kind = DDS_HISTORY_KEEP_LAST;
+  qos.history.depth = 1;
+  qos.destination_order.kind = DDS_DESTINATIONORDER_BY_SOURCE_TIMESTAMP;
+  qos.deadline.deadline = DEADLINE;
+
+  dds_entity_t wr = dds_create_writer(pp, tp, &qos, NULL);
+  CU_ASSERT_FATAL(wr > 0);
+
+  dds_entity_t rd = dds_create_reader (pp, tp, &qos, NULL);
+  CU_ASSERT_FATAL(rd > 0);
+
+  dds_return_t rc = dds_set_status_mask(wr, DDS_PUBLICATION_MATCHED_STATUS);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+
+  uint32_t status = 0;
+  while(!(status & DDS_PUBLICATION_MATCHED_STATUS))
+  {
+    /* Polling sleep. */
+    dds_sleepfor (DDS_MSECS(1));
+
+    CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+    rc = dds_get_status_changes (wr, &status);
+  }
+
+  struct ddsi_domaingv *gvptr = get_domaingv (wr);
+  struct ddsi_xevent *xev = ddsi_qxev_callback(
+    gvptr->xevents,
+    ddsrt_mtime_add_duration(ddsrt_time_monotonic(), (dds_duration_t)(0.5*DEADLINE)),
+    cb,
+    NULL);  //this should sleep the thread that updates the statuses from 0.5*DEADLINE to 1.5*DEADLINE
+  CU_ASSERT_FATAL(xev != NULL);
+
+  Space_Type1 msg1 = { 1, 0, 0 },
```

Comment:
The CI is still failing on macOS, but there is no way of telling where the timing is off ...

There is a very useful function `tprintf` in `test_util` that is a wrapper around printf that first prints a time stamp. Adding calls to it so that the actual duration of the sleeps becomes visible would give a lot more information. (I think it is fine to assume `dds_write` or `check_statuses` take a negligible amount of time.)

Let's do that, check the CI again, and then decide whether to adjust the timing or to change the test to handle timing problems on the execution platform more gracefully.

Reply:
I will give tprintf a try, and see what it spits out.